temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowRunTaskHandler.java

```diff
@@ -267,28 +272,21 @@ public void close() {
     }
   }
 
-  private void processLocalActivityRequests(long startTimeNs) throws InterruptedException {
-    long durationUntilWFTHeartbeatNs =
-        (long)
-            (Durations.toNanos(startedEvent.getWorkflowTaskTimeout())
-                * Config.WORKFLOW_TAK_HEARTBEAT_COEFFICIENT);
-
-    long nextWFTHeartbeatTimeNs = startTimeNs + durationUntilWFTHeartbeatNs;
+  private void processLocalActivityRequests(Deadline wftHeartbeatDeadline)
+      throws InterruptedException {
 
     while (true) {
       List<ExecuteLocalActivityParameters> laRequests =
           workflowStateMachines.takeLocalActivityRequests();
       localActivityTaskCount += laRequests.size();
 
       for (ExecuteLocalActivityParameters laRequest : laRequests) {
-        // TODO(maxim): In the presence of workflow task heartbeat this timeout doesn't make
-        // much sense. I believe we should add ScheduleToStart timeout for the local activities
-        // as well.
-        long maxWaitTimeNs = Math.max(nextWFTHeartbeatTimeNs - System.nanoTime(), 0);
-        laRequest.setScheduleToStartTimeout(Duration.ofNanos(maxWaitTimeNs));
-        boolean accepted = localActivityDispatcher.dispatch(laRequest, localActivityCompletionSink);
-        // TODO this needs to be reworked when we implement a proper scheduleToStart to report a
-        //  Failure in the callback. A proper test is also needed for this scenario.
+        boolean accepted =
+            localActivityDispatcher.dispatch(
+                laRequest, localActivityCompletionSink, wftHeartbeatDeadline);
+        // TODO do we have to fail? if we didn't fit in a potentially tight timeout left until
+        // wftHeartbeatDeadline,
```

Comment:
I don't think you do. In core there is no bound on the local activity queue, just the number of slots.

Reply:
Yeah, it's not really a question, it's just a representation of my thinking. I think we should limit it by one WFT timeout though and fail WFT after to get backpressure all the way up. May be especially important if one worker experiences issues that cause long LA execution and exhausting of LA executor threads.
I will think more about how to improve it. This preserves the existing behavior for now in this regard.